plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java

```diff
@@ -57,15 +59,32 @@ public synchronized AmazonEC2 client() {
             return client;
         }
 
-        this.client = new AmazonEC2Client(buildCredentials(logger, settings), buildConfiguration(logger, settings));
+        AmazonEC2ClientBuilder builder = AmazonEC2ClientBuilder.standard()
+            .withCredentials(buildCredentials(logger, settings))
+            .withClientConfiguration(buildConfiguration(logger, settings));
+
         String endpoint = findEndpoint(logger, settings);
         if (endpoint != null) {
-            client.setEndpoint(endpoint);
+            builder.setEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, buildRegion(endpoint)));
+        } else {
+            logger.debug("No endpoint defined. Using default endpoint [ec2.us-east-1.amazonaws.com].");
         }
 
+        this.client = builder.build();
+
         return this.client;
     }
 
+    protected static String buildRegion(String endpoint) {
+        // We try to get the region from the metadata instance information
+        String region = new InstanceMetadataRegionProvider().getRegion();
```

Comment:
Shouldn't we only be inferring this if the endpoint is *not* set?

Reply:
That should come in another PR related to #27924 IMO.
Here, the goal is to call `new AwsClientBuilder.EndpointConfiguration(endpoint, region)`. As we don't have the region anymore, I need to guess the region either from the instance metadata and if not available from the endpoint with `AwsHostNameUtils.parseRegion(endpoint, null)`.

Make sense?